### ISSUE TITLE
`Normalizer`: list of objects with duplicate object inside leads to `CircularReferenceFoundDuringNormalization`

### ISSUE THREAD

**Issue by boesing** [CuyZ/Valinor#472]
Hey there,

I was trying to switch from JMS to Valinor regarding normalization (as it seems that valinor is way faster and we do not really need most of the JMS magic).

Sadly, when I have this kind of structure where valinor is throwing an exception:

```php
final class Foo
{
     /** @param list<Bar> $collection */
     public function __construct(public readonly array $collection)
     {}
}

final class Bar
{
    public function __construct(public readonly string $baz = '') {}
}

$bar = new Bar();
$anotherBar = new Bar();

$foo = new Foo([$bar, $anotherBar, $bar]);
```

---

We do have a list of objects which is sorted due to its ordering (`list`). We do have to list some entries twice due to feature requirement and thus, we do inject the same object twice (at another position). Its mostly a list of 100+ objects where 3 objects might be duplicated.

I wonder if this is really an issue when normalizing here, I think it would be even fine if there is `anotherBar` referencing `bar` which would throw a circular reference exception as well but I do not see why this would be an issue for normalization.

Was that added intentionally as you ran into a specific case or is that more of a "security mechanism" to prevent projects running into weird infinite loops?

Thanks for this feature, its pretty nice and I'm heading forward to use it.

**Comment by romm** [comment]
Hi @boesing,

> I was trying to switch from JMS to Valinor regarding normalization (as it seems that valinor is way faster and we do not really need most of the JMS magic).
Glad to hear that. Spoiler: I'm working on a compiled version of the normalizer, I've done great progress so far and performances will be even better, like, a lot 🙂 

> Was that added intentionally as you ran into a specific case or is that more of a "security mechanism" to prevent projects running into weird infinite loops?
This specific case was actually a bug, should be solved with #473!

> Thanks for this feature, its pretty nice and I'm heading forward to use it.
Nice! Please don't hesitate to share feedback if you have any. 😉

**Comment by romm** [comment]
And there you go, [released as 1.9.0](https://github.com/CuyZ/Valinor/releases/tag/1.9.0)! 🎉

**Comment by boesing** [comment]
Nice, thanks! Will try that. :)
